monitoring/dns.go

```diff
@@ -40,47 +42,57 @@ func (r *DNSChecker) Name() string { return "dns" }
 
 // Check checks if the DNS servers are responding
 func (r *DNSChecker) Check(ctx context.Context, reporter health.Reporter) {
+	nameservers := r.Nameservers
 	if len(r.Nameservers) == 0 {
-		r.checkWithResolver(ctx, reporter, net.DefaultResolver, "")
-		return
+		clientconfig, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+		if err != nil {
+			reporter.Add(NewProbeFromErr(r.Name(), "failed to load /etc/resolv.conf", err))
+			return
+		}
+
+		nameservers = clientconfig.Servers
 	}
 
-	for _, nameserver := range r.Nameservers {
-		resolver := &net.Resolver{
-			PreferGo: true,
-			Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
-				d := net.Dialer{}
-				return d.DialContext(ctx, "udp", net.JoinHostPort(nameserver, "53"))
-			},
-		}
-		r.checkWithResolver(ctx, reporter, resolver, nameserver)
+	for _, nameserver := range nameservers {
+		r.checkWithResolver(ctx, reporter, nameserver)
 	}
 
 }
 
 func (r *DNSChecker) checkWithResolver(
 	ctx context.Context,
 	reporter health.Reporter,
-	resolver *net.Resolver,
 	nameserver string,
 ) {
 	checkFailed := false
-	for _, question := range r.QuestionA {
-		_, err := resolver.LookupHost(ctx, question)
-		if err != nil {
-			reporter.Add(NewProbeFromErr(r.Name(), errorDetail(question, "A", nameserver), err))
-			checkFailed = true
+	q := new(dns.Msg)
+	q.Id = dns.Id()
+	q.RecursionDesired = true
+	q.Question = make([]dns.Question, 1)
+
+	for questionType, questions := range map[uint16][]string{dns.TypeA: r.QuestionA, dns.TypeNS: r.QuestionNS} {
+		for _, question := range questions {
+			q.Question[0] = dns.Question{question, questionType, dns.ClassINET}
+			in, err := dns.ExchangeContext(ctx, q, fmt.Sprintf("%s:53", nameserver))
+			if err != nil {
+				reporter.Add(NewProbeFromErr(r.Name(), errorDetail(question, dns.TypeToString[questionType], nameserver), err))
+				checkFailed = true
+			}
+			if in.Rcode != dns.RcodeSuccess {
```

Comment:
Will this work if `err != nil` - i.e. will `in` be consistently non-nil?

Reply:
I switched it to continue on error so it doesn't add multiple errors.

Reply:
In that case, I don't see why not `return` early instead of `continue`ing? Do you still want to collect failures from other queries?

Reply:
Yea, I think if multiple queries are configured, it's still relevant to run them.